### PR TITLE
TUIButton clean-up.

### PR DESCRIPTION
**Pull request by avaidyam**
Just a clean-up of `TUIButton` and a few more `NSButton`-corresponding styles. There was a bit of wonky behavior with `TUIControl` as well, but that's fixed now.

### CODE REVIEW

File: lib/UIKit/TUIControl.h [hunk 1]
```diff
@@ -39,7 +40,7 @@ enum {
 	TUIControlEventEditingChanged		= 1 << 17,
 	TUIControlEventEditingDidEnd		= 1 << 18,
 	 */
-	TUIControlEventEditingDidEndOnExit = 1 << 19,
+	TUIControlEventEditingDidEndOnExit	= 1 << 19,
```
Comment (CodaFi) [inline lib/UIKit/TUIControl.h:63]: Please control your white space.

**Comment by CodaFi** [comment]
Ignore the above, I apologize for the extraneous comments.  They'll be turned into diff discussions.

File: lib/UIKit/TUIButton.m [hunk 2]
```diff
@@ -16,246 +16,366 @@
...
+		NSImage *backgroundImage = self.currentBackgroundImage;
+		if(backgroundImage) {
+			[backgroundImage drawInRect:[self backgroundRectForBounds:self.bounds]
+							   fromRect:NSZeroRect operation:NSCompositeSourceOver fraction:1.0];
+		} else {
+			[self.backgroundColor setFill];
+			CGContextFillRect(TUIGraphicsGetCurrentContext(), self.bounds);
+		}
+	} else if(self.buttonType == TUIButtonTypeStandard) {
+		CGFloat level = self.state & TUIControlStateHighlighted ? 0.93 : 0.99;
```
Comment (CodaFi) [inline lib/UIKit/TUIButton.m]: Make this a constant, please.

File: lib/UIKit/TUIButton.m [hunk 3]
```diff
@@ -16,246 +16,366 @@
...
+			} [NSGraphicsContext restoreGraphicsState];
+		}
+	} else if(self.buttonType == TUIButtonTypeBeveled) {
+		NSBezierPath *path = [NSBezierPath bezierPathWithRect:self.bounds];
+		NSGradient *bevel = [[NSGradient alloc] initWithColorsAndLocations:
+							 [NSColor colorWithCalibratedWhite:0.95 alpha:0.5], 0.0,
+							 [NSColor colorWithCalibratedWhite:0.90 alpha:0.5], 0.5,
+							 [NSColor colorWithCalibratedWhite:0.85 alpha:0.5], 0.5,
+							 [NSColor colorWithCalibratedWhite:0.84 alpha:0.5], 1.0, nil];
+
```
Comment (CodaFi) [inline lib/UIKit/TUIButton.m:196]: Constants, please.  Or better yet, avoid NSGradient and make two static CGFloat matrices pass to `CGGradientCreateWithColorComponents()`.

File: lib/UIKit/TUITextView.m [hunk 4]
```diff
@@ -811,7 +811,7 @@ TUIViewDrawRect TUITextViewStandardFrame(void)
 		// black outline
 		NSRect blackOutlineFrame = NSMakeRect(0.0, 1.0, bounds.size.width, bounds.size.height-2.0);
 		NSGradient *gradient = nil;
-		if([NSApp isActive]) {
+		if(![NSApp isActive]) {
 			gradient = [[NSGradient alloc] initWithStartingColor:[NSColor colorWithDeviceWhite:0.6 alpha:1.0] endingColor:[NSColor colorWithDeviceWhite:0.7 alpha:1.0]];
 		} else {
 			gradient = [[NSGradient alloc] initWithStartingColor:[NSColor colorWithDeviceWhite:0.55 alpha:1.0] endingColor:[NSColor colorWithDeviceWhite:0.558 alpha:1.0]];
```
Comment (CodaFi) [inline lib/UIKit/TUITextView.m:817]: This should be in a separate PR.

File: lib/UIKit/TUIButton.h [hunk 5]
```diff
@@ -14,11 +14,6 @@
  limitations under the License.
  */
 
-/*
- TUIButton adopts some baggage from UIButton, probably a mistake
- this class is likely to change (be simplified) radically.
- */
```
Comment (CodaFi) [inline lib/UIKit/TUIButton.h:20]: (Oh dear, sweet irony).  Perhaps in the interests of documenting the class, you could replace this with a short description of TUIButton.

File: lib/UIKit/TUIButton.m [hunk 6]
```diff
@@ -16,246 +16,366 @@
...
+							 [NSColor colorWithCalibratedWhite:0.85 alpha:0.5], 0.5,
+							 [NSColor colorWithCalibratedWhite:0.84 alpha:0.5], 1.0, nil];
+		
+		[self.tintColor set];
+		[path fill];
+		
+		[NSGraphicsContext saveGraphicsState];
+		[[NSShadow shadowWithRadius:1.0
+							 offset:NSMakeSize(0, -1.0)
+							  color:[NSColor colorWithCalibratedWhite:.863 alpha:.75]] set];
```
Comment (CodaFi) [inline lib/UIKit/TUIButton.m]: If we're gonna get anal about our magic numbers, it would be best to declare the ...calibratedWhite: arg 0.86f, and the alpha arg as 0.75f if only to be explicit. Or avoid this altogether. Make these constants.

File: lib/UIKit/TUIButton.h (hunk 5 above)
Comment (avaidyam) [inline lib/UIKit/TUIButton.h:20]: I was going to do that after code review ;D

File: lib/UIKit/TUITextView.m (hunk 4 above)
Comment (avaidyam) [inline lib/UIKit/TUITextView.m:817]: Oops.

File: lib/UIKit/TUIButton.m (hunk 3 above)
Comment (avaidyam) [inline lib/UIKit/TUIButton.m:196]: I prefer Cocoaland, thank you ver much! Much more readable, but I will make them constants.

**Comment by avaidyam** [comment]
In regards to the `super` calls, there was nothing in the method body to begin with +____+ I propose we remove them, but that's another PR for later.

File: lib/UIKit/TUIButton.m (hunk 6 above)
Comment (avaidyam) [inline lib/UIKit/TUIButton.m]: That would be waaayyy too many constants.... how do you suppose I go about adding 50 constants for colors! :laughing: But yeah, I don't like constants when there's too many (or they're for complex drawing). Another benefit of leaving things like this, is that unlike `UIButton`, `TUIButton` is _subclass-safe_! :yum:

Comment (CodaFi) [inline lib/UIKit/TUIButton.m]: Eliminate the colors!  You made this class overly complicated.  All of the buttons types you made can be produced on demand by the programmer by providing a `TUIViewDrawRect` block.  While I do love the tweaks and changes you made to the class, adding the various options that you did created complexity which _you_ have to either deal with, or gut.   

It really makes me start to realize how AppKit got to be a bloated mess in the first place (no offense intended, just my observation).

Comment (avaidyam) [inline lib/UIKit/TUIButton.m]: Alright, but I've only added the simpler ones >_> Sure, a few of them could be removed, but ones like `Standard`, `Minimal`, or `Inline` make sense to keep around...

**Comment by avaidyam** [comment]
By the way, @CodaFi , how are you "assigned" to an Issue in the tracker? .___.

**Comment by CodaFi** [comment]
> Alright, but I've only added the simpler ones >_> Sure, a few of them could be removed, but ones like Standard, Minimal, or Inline make sense to keep around...

Maybe, if we wanted it to be a suitable replacement for AppKit, but I believe it was @joshaber who said that TwUI was about "Making the best damn layer-backed framework around."  As @jwilling suggested, you should seriously consider making a repository of all of your work.  What you're doing isn't bad, in fact, it's really quite brilliant, but you have to understand that the framework isn't meant to be a catch-all, nor an all-around, all-out framework.  The framework is meant to be only the **essentials**, which can then be tweaked and subclasses (as you have done many a time), to make innovative controls and new ways of interacting with our computers.

That said, I'm not okay with this merging until the class is either finalized or simplified.

**Comment by avaidyam** [comment]
@CodaFi Sure, I can definitely remove some of the stuff in there. Half the reason I bloat whatever PRs I do have, is because I try to maintain API consistency for new developers. To be honest, I do think, even on 10.8, AppKit is 20-year old bloat that needs to rot. UIKit is fantastic, but too locked down. So I'm trying to bring a middle-approach. 

That being said, I do understand what you all mean, and I'm perfectly fine removing some of the stuff in there that might bloat the class. I'll do that after I figure out why my scroll bars are having seizures (another PR >_>).

**Comment by CodaFi** [comment]
Same here.  Merge upstream, and redo the example before I continue doing review.

**Comment by avaidyam** [comment]
@CodaFi Here's the big update! It's more or less finished on my end.

**Comment by avaidyam** [comment]
@CodaFi I've cleaned up a lot and merged upstream. :mailbox_with_mail:

**Comment by jwilling** [comment]
First of all, I'm not sure if this is intentional or not but this looks horrible.


Also, try resizing the demo app. I don't know if this was introduced in this pull request, but it's bad.

**Comment by avaidyam** [comment]
@jwilling HOLY JESUS WHAT THE HECK. Alright, something is seriously up with this. I'm looking into it ASAP. :sheep:

**Comment by avaidyam** [comment]
@jwilling The first issue is caused by `NSShadow` not working nicely with Retina displays. I have a fix for that as well.

**Comment by jwilling** [comment]
@galaxas0 Please introduce that fix in a separate pull request.

**Comment by avaidyam** [comment]
@jwilling Got it, and I'm working on it!

## 

Aditya.

**Comment by avaidyam** [comment]
@jwilling I've fixed the issues with the example project. :mailbox_with_mail:

**Comment by jwilling** [comment]
What exactly are these buttons supposed to look like? This? I don't know what style of control you're trying to replicate but it does not look good.

**Comment by jwilling** [comment]
There are far too many changes in this PR. Please split a significant portion of them up as appropriate. Attempting to review a pull request of this size can only lead to errors that slip through accidentally. Give this a bump when you think it's ready for review again.

**Comment by avaidyam** [comment]
@jwilling Here's how it's supposed to look on non-Retina:
- window not key: http://i50.tinypic.com/2mxhoq8.png
- standard: http://i50.tinypic.com/nw4rk.png
- highlighted or selected: http://i46.tinypic.com/9j3892.png

**Comment by avaidyam** [comment]
@jwilling I'll separate the example app modifications and the actual button update. Clearly, the issues you're seeing are Retina only, and since I won't be getting a Retina MBP for another 6 months, I can't really attempt to fix them! Sorry! :cake:

**Comment by jwilling** [comment]
@galaxas0 More than just that. For example, the shadow additions should be in a PR.

Also, you can test HiDPI mode on a standard-resolution computer. http://osxdaily.com/2012/01/12/enable-hidpi-mode-in-mac-os-x-lion/

**Comment by avaidyam** [comment]
@jwilling Alright, I'll add that too, in a separate PR (god, I have so many of these now!)

I've tried that before, and with a 15 inch screen, things get _seriously_ unusable in HiDPI. Programming? More like have-fun-having-more-than-one-window-open. >_>

**Comment by jwilling** [comment]
@galaxas0 In any case it can help you test your presumed-working code before you release it, perhaps.

**Comment by avaidyam** [comment]
@jwilling Sure, I can use it for at least 2 minutes to test graphics stuff… :)

**Comment by avaidyam** [comment]
@jwilling @CodaFi I've done some testing using HiDPI mode on a new Mac Mini, and realized several flaws:
- The button style does not have OS X version consistency: 10.6, 10.7, and 10.8+ all have different button drawing.
- The button style does not have User-wide consistency: the user will notice quirky buttons easily.
- The button style does not have proper HiDPI capable drawing (`NSShadow` can go die in a ditch :cry:).

After some investigation, I've determined that AppKit's demon-child `NSCell` is actually quite useful for us: We can use these little drawing blocks on top of our amazing `CALayer` backed `TUIView`s easily and keep OS and User consistency.

`NSButtonCell` uses the `CoreUI` framework introduced in 10.5 to draw all its controls perfectly, and can do everything for us: so why do it ourselves? Now `TUIButton` has a shared button renderer (an `NSButtonCell`) whose bezel style is swapped out and is redrawn for each button.

Now, `TUIButton` is indiscernible from a standard `NSButton`! I've yet to figure out how to pulse the button though: my idea is to use a `CAAnimation` to swap button colors, but there's no way to find the "darker" button color in the AppKit pulse. AppKit also uses a special `NSHeartBeat` thread for these animations (I believe).

Can we reopen this now? :+1:

**Comment by CodaFi** [comment]
Make a new PR for NSCell'd buttons.  I'm not reopening this and merging the previous stuff.

**Comment by avaidyam** [comment]
@CodaFi I'd like to point out that they're not `NSCell'd`. :wink: They use the class's shared `NSButtonCell` as a renderer for the background. It's global for all button instances. I'll create a new PR though.
